rust_src/src/filelock.rs

```diff
@@ -2,13 +2,277 @@
+const MAX_LOCK_INFO: usize = 8 * 1024;
+
+/// The lock info as parsed from a string like `USER@HOST.PID:BOOT_TIME`.
+#[derive(PartialEq, Eq, Debug)]
+struct LockInfo {
+    /// The user that owns the lock.
+    user: String,
+    /// The host on which the file was locked.
+    host: String,
+    /// The ID of the process that locked the file.
+    pid: i32,
+    /// Boot time of the system that locked the file, as Unix epoch seconds.
+    boot_time: Option<i64>,
+}
+
+impl LockInfo {
+    /// Parses a string like `USER@HOST.PID:BOOT_TIME` into a [`LockInfo`].
+    /// Returns [`None`] if the parse fails.
+    fn parse(data: &str) -> Option<LockInfo> {
+        // Treat "\357\200\242" (U+F022 in UTF-8) as if it were ":" (Bug#24656).
+        // This works around a bug in the Linux CIFS kernel client, which can
+        // mistakenly transliterate ':' to U+F022 in symlink contents.
+        // See https://bugzilla.redhat.com/show_bug.cgi?id=1384153
+        let boot_time_sep = |c| c == ':' || c == '\u{F022}';
+
+        // The USER is everything before the last @.
+        let (user, rest) = data.split_at(data.rfind('@')?);
+        let user = user.to_string();
+
+        // The HOST is everything after the '@' to the last '.'.
+        let (at_host, rest) = rest.split_at(rest.rfind('.')?);
+        let host = at_host[1..].to_string();
+
+        // The PID is everything from the last '.' to the ':' or equivalent.
+        // NOTE: Unlike GNU Emacs, we do not tolerate negative PID values
+        // or over/underflow.
+        let (pid_str, boot_time_str) =
+            rest.split_at(rest.find(boot_time_sep).unwrap_or(rest.len()));
+        let pid = pid_str[1..].parse::<i32>().ok().filter(|pid| pid > &0)?;
+
+        // After the ':' or equivalent, if there is one, comes the boot time.
+        let boot_time = if boot_time_str.is_empty() {
+            None
+        } else {
+            // Invalid boot time results in a failed parse of the LockInfo
+            // rather than just boot_time = None.
+            // NOTE: Unlike GNU Emacs, this also applies to over/underflow.
+            // NOTE: Negative value is allowed.
+            Some(
+                boot_time_str
+                    .trim_start_matches(boot_time_sep)
+                    .parse()
+                    .ok()?,
+            )
+        };
+
+        Some(LockInfo {
+            user,
+            host,
+            pid,
+            boot_time,
+        })
+    }
+}
+
+/// The current locking state of a file.
+#[derive(PartialEq, Eq, Debug)]
+enum LockState {
+    /// The file is not locked.
+    NotLocked,
+    /// The file is locked by this Emacs process.
+    LockedByUs,
+    /// The file is locked by some other Emacs process identified
+    /// by [`LockInfo`].
+    LockedBy(LockInfo),
+}
+
+/// Opens a file in `O_NOFOLLOW` mode. Returns [`None`] if the open failed
+/// because the `path` refers to a symbolic link.
+#[cfg(unix)]
```

Comment:
Can you add stubs for windows please? That when someone wants Windows to work we have all of the steps laid out for them.